### DISCUSSION TITLE
Buildpacks default process behavior is confusing (We should use `bash -c` as a process type)

### DISCUSSION THREAD

**Discussion by schneems** [heroku/buildpacks#15] [Ideas]
## Expected

When I build an image with a CNB then `docker run ... bash` that I will get an interactive shell to the image

## Actual

Example, if you follow this part of the tutorial https://github.com/heroku/buildpacks/blob/71afce6c7a4df02ff3d48830dc1c1fa8f574ccf7/TUTORIAL.md#configuring-multiple-languages but delete the procfile buildpack, and then try to run ` docker run -it --rm my-python-nodejs-image bash` it will boot a webserver instead of giving you an interactive session.

## Analysis

(paraphrasing) The behavior of the CNB spec says that the values passed to the container will be passed to the `docker run` instance will be added as arguments to the default command. So if the default command is `npm run start` then calling `docker run ... bash` will effectively call `npm run start bash` which is not what we want.

The reason the procfile buildpack doesn't have this problem is because it uses `bash -C` as the command and then passes in the rest as arguments. https://github.com/heroku/buildpacks-procfile/blob/58ca67cc03b78009fb821f26d2b8c39b55659478/src/launch.rs#L19 so that means if you give `docker run` the arg `bash` it will run `bash -C bash` which does what we expect

## Workarounds

The app user can workaround this issue by manually invoking the launcher as the `--entrypoint`, but this requires the end user to have knowledge of how the launcher works, and why this is needed. When the system does not behave as the user would expect, there is no hint for the application developer to look towards using the `--entrypoint` flag in docker. For example:

```
$    docker run -it --rm --entrypoint='/cnb/lifecycle/launcher' "$imagename" bash
```

## Into the future 

This would be improved by "system buildpacks" as described in this RFC https://github.com/buildpacks/rfcs/blob/ce8991b3d739f763b8356ae8696ae582f18c4af6/text/0101-system-buildpacks-in-builder.md?plain=1#L3 however this is accepted and not yet implemented. This would allow the builder to declare that the Procfile buildpack should always execute after any build made with the builder.

However, there will still be edge cases, for example, when someone does not have a Procfile in the root of their application.

## Strategy 1 - `bash -C` all the things

All Heroku buildpacks could adopt the same `bash -C` strategy that the Procfile buildpack uses. We could recommend this strategy in documentation as well.

## Strategy 2 - Change the default behavior upstream

The default behavior of `docker run` is to run the default command and pass any arguments to `docker run` to that default command instead of replacing it. We could try to be clever and optionally replace the command based on what's being passed in. Alternatively we could remove this "run default command" behavior so the user would be forced to run `docker run ... web` if they want to boot the web process. Or we could change it so that `docker run` with no arguments runs the default command and any arguments are expected to be a full command replacement, rather than only arguments to that command.

**Comment by schneems** [comment]
I think we should adopt strategy 1 (bash -C our buildpacks) as a short term solution while socializing and exploring strategy 2 (changing upstream behavior).

**Reply by schneems** [reply]
I was also thinking: It could be possible to build both a rust-specific CNB suite of lining tools as well as a higher level "CNB" linting tools, for example: Build an image with a buildpack, run linting tools, it could error/warn if it finds a process without `bash -c` (that is ignorable/configurable).

**Comment by edmorley** [comment]
This recently came up in:
https://github.com/heroku/buildpacks-nodejs/issues/906

**Comment by nickhammond** [comment]
This would be great! Just stumbled on this using the ruby buildpack when it said it couldn't find ruby. I was attempting to run `bin/rails console` within my container.

```
/usr/bin/env: 'ruby': No such file or directory
```

I realized I could probably utilize the existing entrypoint process which worked but it's definitely not intuitive.

```
/cnb/process/web "bin/rails console"
```

**Comment by dzuelke** [comment]
Wrapping in `bash -c` has the problem that suddenly, argument handling becomes different from what you have with a vanilla Docker image, and I don't think we want or need another subtle difference as a result of magic that I don't think is particularly helpful anyway - automatically starting `web` doesn't do anything for the user if the user doesn't specify a port mapping and matching port env var, so the user already needs some knowledge about process types, defaults, ands what `web` needs.

I think we should instead not have default process types at all, and show a "help screen" if the Docker image is run without arguments. Right now, `launcher` just says that a command is required, but that can easily be changed to be more instructive (tell people about `bash`, but also list the known process type's binaries): https://github.com/buildpacks/lifecycle/blob/92d9833a465c8dd0949f86b4ee60f6bfc86e02c9/launch/process.go#L114

And as an added bonus, the process type binaries each could maybe have a help screen that gives users instructions on what's needed to launch that process type (so for `web` it would say to have a port mapping, and set the `PORT` env var).

**Reply by colincasey** [reply]
Using a detailed and instructive `--help` message as the default process is a pretty great idea @dzuelke.

**Reply by nickhammond** [reply]
The `inspect-image` command might also be a helpful place for this. It's mentioned in the docs as a way to figure out what processes are defined for an image. 

`pack inspect-image image-name`

```
Processes:
  TYPE                 SHELL        COMMAND        ARGS                                                 WORK DIR
  web (default)                     bash           -c ./bin/docker-entrypoint ./bin/rails server        /workspace
```

- Docs for inspect-image - https://buildpacks.io/docs/for-app-developers/how-to/build-outputs/specify-launch-process/
- launcher specs - https://github.com/buildpacks/spec/blob/main/platform.md#launcher)

**Reply by schneems** [reply]
Interesting proposal. I see this as two parts

> Right now, launcher just says that a command is required, but that can easily be changed to be more instructive (tell people about bash, but also list the known process type's binaries)

100% agree. Seems good to improve the messaging when we know what someone is wanting to do. I've not hit this error message yet personally, but trust others have. Great suggestion. Yes, and...

> I think we should instead not have default process types at all

Ruby customers expect to be able to deploy to Heroku and it just work i.e. `rails new && git push heroku`, which involves setting default process types. We're the community of "convention over configuration." I don't feel comfortable dropping default process type from the Ruby buildpack. This is a core value add for buildpacks, at least in the Ruby ecosystem.

So, I support enhancing the launcher message. However, I don't agree that we would be better off eliminating default process types with the current CNBs.

We could possibly propose a change to the spec like removing the behavior upstream so `docker run` with no args ran nothing except printed that message. And then `docker run default`  booted whatever web process is marked as "default." That could work, but would break anyone relying on current behavior (like the paketo getting started guide https://paketo.io/docs/) and anyone using `default` as a process type. The path to getting a change like that merged is not certain.

**Reply by prullmann** [reply]
Being fairly new to CNBs, I recently got confused by this issue.

I would have expected that
`docker run ...` would run the web server, but
`docker run ... bash` would give me a shell.

To run the web server with arguments, the spec says I should not give a cmd, just args. But you can't omit the command with docker, the first arg will just be the cmd, which is a bit confusing.

So when running the web server with arguments I would find it much more intuitive to mention the command as well, like `docker run ... web -p 80`.
This would also be more in line with what is in a Procfile.

**Reply by schneems** [reply]
Thanks for the user experience @prullmann. Building on David's idea of a "help" screen. I think we could do both print helpful info and not lose the existing default process support. Lifecycle is already inserted between `docker` and the command that's being run:

```
$ docker inspect my-image-name | grep '"Entrypoint": \[' -A2
            "Entrypoint": [
                "/cnb/process/web"
            ],
```

We could theoretically echo out the command it's about to run. So then if someone does a `docker run ... bash` and they see 

> Running "bin/rails server" with arguments "bash" 

They would at least have a hint as to what's happening under the covers. It could be extended to print a list of other known 

It could also special case receiving only one argument and it being "bash" to print out a switchable help message. Like:

> Running "bin/rails server" with arguments "bash"
> INFO: If you are intending to run a `docker run ... bash` interactive console use `docker run --entrypoint =... bash`. 
> Silence this message by setting `export PACK_SILENCE_DOCKER_BASH_MESSAGE=1` in your shell initialization script.

Another idea: Introduce a `pack bash` command for interactive access. It would be pretty cool if buildpacks could also define default things like port mappings and env vars and directory mappings. But that would be a much larger undertaking. We would need to align on scope, security, and interfaces, and those are all ripe for bikeshedding

**Comment by dzuelke** [comment]
I implemented a proof of concept here: https://github.com/heroku/buildpacks-php/pull/229

See that PR for details and "screenshots" of what it looks like, but basically, a new default process type named `usage` shows a help screen with information on how to invoke the image, allows `docker run <image-name> help <process-type>`, and also lets users run arbitrary commands simply via `docker run --rm -ti <image-name> bash`, or `docker run --rm <image-name> bash -c 'echo "hello world"'` and so forth - no special argument parsing rules etc.

It can fetch a help screen text file from `/layers/<buildpack>/usage_help/<process-type>.txt` for a process type; if that is not present, it shows some generic help screen.

Having this be the default process type is much more straightforward than cramming it into `launcher`, I think. Also, I realized a while ago that having `web` be the default process type really isn't very useful, because just `docker run`ning an image might start the app (if a web server binds to a default port in the absence of `PORT` in the environment), but it won't be accessible without a `-p` for port mapping, so showing a help screen that tells the user to do `docker run --rm --entrypoint web -p 8080:8080 -e PORT=8080 <image-name>` makes for a much better "out of the box" UX (also in a "teach a person to fish" sense).

Right now it's implemented inside an existing buildpack, but as its own buildpack, it could always run last in a builder's various groups, and probably skip execution if it detects that it's not being run in a local `pack build` (so that e.g. a Heroku user doesn't end up with a `usage` process type in their process table).
